File: frontend/packages/types/src/local-event.ts
/** Local events / city moments (TICKET-505). */

import type { FeedNeighborhoodSummary } from "./neighborhood";
import type { EventReadinessFields } from "./event-readiness";
import type { PartnerStatus } from "./partner";

export interface LocalEventOrganization {
  id: string;
  slug: string;
  name: string;
  city: string;
  logo_url: string | null;
  is_verified?: boolean;
  is_partner?: boolean;
  partner_status?: PartnerStatus | null;
  created_at?: string | null;
}

export interface LocalEvent {
  id: string;
  organization_id: string | null;
  title: string;
  description: string | null;
  event_type: string | null;
  city: string;
  district: string | null;
  starts_at: string;
  ends_at: string | null;
  timezone: string;
  location_name: string;
  address: string | null;
  latitude: number | null;
  longitude: number | null;
  cover_image_url: string | null;
  moderation_status: string;
  is_cancelled: boolean;
  interested_by_me: boolean;
  /** Présent sur les réponses API ; optionnel dans les mocks de tests. */
  interest_count?: number;
  organization: LocalEventOrganization | null;
  neighborhood_summary?: FeedNeighborhoodSummary | null;
  created_at: string;
}

export interface LocalEventListResponse {
  items: LocalEvent[];
  total: number;
  page: number;
  page_size: number;
}

export interface EventInterestToggleResponse {
  event_id: string;
  interested: boolean;
  interest_count: number;
}

export interface FeedEventMeta {
  local_event_id: string;
  starts_at: string;
  ends_at: string | null;
  location_name: string;
  district: string | null;
  event_type: string | null;
  interested_by_me?: boolean;
}

export interface PartnerEventsParams {
  upcoming_only?: boolean;
  limit?: number;
  offset?: number;
}

/** Organization self-service (TICKET-505 / WEB-PARTNERS-08C). */
export interface LocalEventManagement extends LocalEvent {
  rejection_reason?: string | null;
  readiness: EventReadinessFields;
}

export interface LocalEventManagementListResponse {
  items: LocalEventManagement[];
  total: number;
  page: number;
  page_size: number;
}

export interface LocalEventCreatePayload {
  organization_id: string;
  title: string;
  description?: string | null;
  event_type?: string | null;
  city: string;
  district?: string | null;
  starts_at: string;
  ends_at?: string | null;
  timezone?: string;
  location_name: string;
  address?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  cover_image_url?: string | null;
}

export interface LocalEventUpdatePayload {
  title?: string;
  description?: string | null;
  event_type?: string | null;
  district?: string | null;
  starts_at?: string;
  ends_at?: string | null;
  location_name?: string;
  address?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  cover_image_url?: string | null;
}

export interface LocalEventManagementListParams {
  organization_id?: string;
  page?: number;
  page_size?: number;
}

File: backend/app/schemas/local_event.py
"""Local events API schemas (TICKET-505)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.core.local_event_constants import (
    LOCAL_EVENT_LIST_PAGE_SIZE_DEFAULT,
    MVP_LOCAL_EVENT_TYPES,
)
from app.schemas.event_readiness import EventReadinessFields
from app.schemas.neighborhood import FeedNeighborhoodSummary


class LocalEventOrganizationSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    name: str
    city: str
    logo_url: str | None = None
    is_verified: bool = False
    is_partner: bool = False
    partner_status: str | None = None
    created_at: datetime | None = None


class LocalEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: uuid.UUID | None
    title: str
    description: str | None
    event_type: str | None
    city: str
    district: str | None
    starts_at: datetime
    ends_at: datetime | None
    timezone: str
    location_name: str
    address: str | None
    latitude: float | None = None
    longitude: float | None = None
    cover_image_url: str | None
    moderation_status: str
    is_cancelled: bool
    interested_by_me: bool = False
    interest_count: int = 0
    organization: LocalEventOrganizationSummary | None = None
    neighborhood_summary: FeedNeighborhoodSummary | None = None
    created_at: datetime


class LocalEventListResponse(BaseModel):
    items: list[LocalEventResponse]
    total: int
    page: int = 1
    page_size: int = LOCAL_EVENT_LIST_PAGE_SIZE_DEFAULT


class LocalEventCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    organization_id: uuid.UUID
    title: str = Field(min_length=3, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    event_type: str | None = Field(default=None, max_length=64)
    city: str = Field(min_length=2, max_length=128)
    district: str | None = Field(default=None, max_length=128)
    starts_at: datetime
    ends_at: datetime | None = None
    timezone: str = Field(default="Europe/Paris", max_length=64)
    location_name: str = Field(min_length=2, max_length=200)
    address: str | None = Field(default=None, max_length=300)
    latitude: float | None = None
    longitude: float | None = None
    cover_image_url: str | None = Field(default=None, max_length=500)


class LocalEventUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=3, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    event_type: str | None = Field(default=None, max_length=64)
    district: str | None = Field(default=None, max_length=128)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    location_name: str | None = Field(default=None, min_length=2, max_length=200)
    address: str | None = Field(default=None, max_length=300)
    latitude: float | None = None
    longitude: float | None = None
    cover_image_url: str | None = Field(default=None, max_length=500)


class LocalEventManagementResponse(LocalEventResponse):
    rejection_reason: str | None = None
    readiness: EventReadinessFields


class LocalEventManagementListResponse(BaseModel):
    items: list[LocalEventManagementResponse]
    total: int
    page: int
    page_size: int


class LocalEventRejectRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=3, max_length=500)


class EventInterestToggleResponse(BaseModel):
    event_id: uuid.UUID
    interested: bool
    interest_count: int = 0


class LocalEventTypeListResponse(BaseModel):
    items: list[str] = Field(default_factory=lambda: list(MVP_LOCAL_EVENT_TYPES))
